core/domain/exp_domain.py

```diff
@@ -2101,6 +2110,71 @@ def _convert_states_v25_dict_to_v26_dict(cls, states_dict):
 
         return states_dict
 
+    @classmethod
+    def _convert_states_v26_dict_to_v27_dict(cls, states_dict):
+        """Converts from version 26 to 27. Version 27 adds content_translations
+        dict to the state, which will allow translators to add translation
+        script for the state contents.
+
+        NOTE: This migration will also filter out the content_id from
+        content_ids_to_audio_translations such that the state passes the new
+        validation check safely.
+
+        Earlier state validation use to check that the set of all content ids
+        present within the state is sub-set of the
+        content_ids_to_audio_translations keys, but the new validation will
+        check whether both are equal.
+
+         Args:
+            states_dict: dict. A dict where each key-value pair represents,
+                respectively, a state name and a dict used to initialize a
+                State domain object.
+
+        Returns:
+            dict. The converted states_dict.
+        """
+        for _, state_dict in states_dict.iteritems():
+            state_content_id_list = []
+
+            # Add state card's content id into the state_content_id_list.
+            state_content_id_list.append(state_dict['content']['content_id'])
+
+            # Add answer_groups content id into the state_content_id_list.
+            for answer_group in state_dict['interaction']['answer_groups']:
+                answer_feedback = answer_group['outcome']['feedback']
+                state_content_id_list.append(answer_feedback['content_id'])
+
+            # If present add default_outcome content id into
+            # state_content_id_list.
+            default_outcome = state_dict['interaction']['default_outcome']
+            if default_outcome:
+                state_content_id_list.append(
+                    default_outcome['feedback']['content_id'])
+
+            # Add hints content id into state_content_id_list.
+            for hint in state_dict['interaction']['hints']:
+                state_content_id_list.append(hint['hint_content']['content_id'])
+
+            # If present add solution content id into state_content_id_list.
+            solution = state_dict['interaction']['solution']
+            if solution:
+                state_content_id_list.append(
+                    solution['explanation']['content_id'])
+
+            # Filter content_ids_to_audio_translations with unwanted content id.
+            citat = state_dict['content_ids_to_audio_translations']
+            extra_content_ids_in_citat = (
```

Comment:
Why would there be extra content IDs?

Is there potential for data loss here? If we don't think there is, could we validate that?

Reply:
OK -- please add comments in the code to explain this and explain clearly the reasoning behind why no (important) data would be lost here.

Reply:
I have aded the reason for "no data loss" and the reson for why these content ids (can be) present within the content_ids_to_audio_translations is provided in the docstring of the migration function.